File: src/main/java/com/everchanging/hierarchy/entity/EmployeeEntity.java
package com.everchanging.hierarchy.entity;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.OneToOne;
import javax.persistence.Table;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "employees")
public class EmployeeEntity {
    @Id
    private String name;

    @OneToOne
    private EmployeeEntity supervisor;
}

File: src/main/kotlin/com/everchanging/hierarchy/entity/EmployeeEntity.kt
package com.everchanging.hierarchy.entity

import javax.persistence.*

@Entity
@Table(name = "employees")
data class EmployeeEntity(
        @Id
        var name: String?,
        @OneToOne(
                fetch = FetchType.LAZY // do not load all upper level employees
        )
        private var supervisor: EmployeeEntity?) {
    constructor() : this(name = null, supervisor = null)
}